### PR TITLE
add option for 'private field initialization' lint

### LINKED ISSUE #81447: Incorrect private field diagnostic when `self` is passed
I tried this code ([playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=74b6cc513cc3e8e131b1b31805d0240e)):

```rust
mod some_module {
    pub struct Test<T: ?Sized>(T);
}

use some_module::Test;

struct TestBuilder;

impl TestBuilder {
    fn build(self) -> Test {
        Test(self)
    }
}
```

I expected to receive the "cannot initialize a tuple struct which contains private fields" lint.

Instead, this happened:

```
error[E0423]: expected function, tuple struct or tuple variant, found struct `Test`
  --> src/lib.rs:11:9
   |
11 |         Test(self)
   |         ^^^^------
   |         |
   |         help: try calling `Test` as a method: `self.Test()`
```

it appears the "method on self" lint takes priority over the "private field initialization" lint when it likely shouldn't, at least in this scenario.

### Meta

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (d9a105fdd 2020-11-23)
binary: rustc
commit-hash: d9a105fdd46c926ae606777a46dd90e5b838f92f
commit-date: 2020-11-23
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

(also happens on stable, see the playground link provided)

</p>
</details>

### PR DESCRIPTION
**Pull request by henryboisdequin**
Fixes #81447 by adding an option to emit the 'private field initialization' lint before emitting the 'method on self' lint

### CODE REVIEW

File: src/test/ui/self/suggest-self-2.stderr [hunk 1]
```diff
@@ -2,25 +2,19 @@ error[E0425]: cannot find function `bar` in this scope
   --> $DIR/suggest-self-2.rs:5:9
    |
 LL |         bar(self);
-   |         ^^^------
-   |         |
-   |         help: try calling `bar` as a method: `self.bar()`
```
Comment (henryboisdequin) [inline src/test/ui/self/suggest-self-2.stderr:7]: Hmm, I think I need to add another check

Comment (henryboisdequin) [reply src/test/ui/self/suggest-self-2.stderr:7]: Fixed

**Comment by henryboisdequin** [comment]
Friendly ping @lcnr or maybe someone else wants to review it?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 2]
```diff
@@ -428,12 +428,33 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
                 }
 
-                err.span_suggestion(
-                    call_span,
-                    &format!("try calling `{}` as a method", ident),
-                    format!("self.{}({})", path_str, args_snippet),
-                    Applicability::MachineApplicable,
-                );
+                if source.is_call() {
+                    if let Some(val) = res {
```
Comment (lcnr) [inline compiler/rustc_resolve/src/late/diagnostics.rs:432]:
```suggestion
                    if let Some(res) = res.filter(|_| source.is_call()) {
```
to not duplicate the else branch

**Comment by lcnr** [comment]
don't know much about error reporting during name resolution

r? @petrochenkov maybe

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 3]
```diff
@@ -428,12 +428,24 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
                     }
                 }
 
-                err.span_suggestion(
-                    call_span,
-                    &format!("try calling `{}` as a method", ident),
-                    format!("self.{}({})", path_str, args_snippet),
-                    Applicability::MachineApplicable,
-                );
+                if let Some(val) = res.filter(|_| source.is_call()) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:431]: How does this work?
Looks like `call_has_self_arg` above ensures that `source.is_call()` is true.
So the condition is just `if let Some(val) = res`, i.e. "some resolution found, but it doesn't meet restrictions", which seems irrelevant to the issue that you are trying to solve.

File: src/test/ui/self/suggest-self-2.rs [hunk 4]
```diff
@@ -4,21 +4,18 @@ impl Foo {
     fn foo(&self) {
         bar(self);
         //~^ ERROR cannot find function `bar` in this scope
-        //~| HELP try calling `bar` as a method
```
Comment (petrochenkov) [inline src/test/ui/self/suggest-self-2.rs:7]: Could you avoid unnecessary changes (this test, formatting in `compiler/rustc_resolve/src/late/diagnostics.rs`)?

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 3 above)
Comment (henryboisdequin) [reply compiler/rustc_resolve/src/late/diagnostics.rs:431]: @petrochenkov, Thanks for your quick review. Yes, you're right. I was thinking that the  `smart_resolve_context_dependent_help` function would do all the checks. Does that mean the `if let Some(val) = res` is irrelevant, because I do need that `val` in order to call `smart_resolve_context_dependent_help`? Also, for the test stderr in `compiler/rustc_resolve/src/late/diagnostics.rs`, the help messages of `try calling bar as a method` are still there. Doesn't that mean that, in some cases, `source.is_call` is false?

Comment (petrochenkov) [reply compiler/rustc_resolve/src/late/diagnostics.rs:431]: >Does that mean the `if let Some(val) = res` is irrelevant, because I do need that `val` in order to call `smart_resolve_context_dependent_help`?

It is irrelevant because the "try calling `{}` as a method" heuristic may work regardless of whether there's no resolution or some inappropriate resolution is found.

>Also, for the test stderr in `src/test/ui/self/suggest-self-2.rs`, the help messages of `try calling bar as a method` are still there. Doesn't that mean that, in some cases, `source.is_call` is false?

It only means that `res` is `None` in that case.
`is_call` should always be true at this point due to the `call_has_self_arg` above.

**Comment by henryboisdequin** [comment]
@petrochenkov Thanks for your comments, I believe that I have addressed all of them.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 5]
```diff
@@ -428,12 +428,24 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                        &fallback_label,
+                    );
+                } else {
+                    err.span_suggestion(
+                        call_span,
+                        &format!("try calling `{}` as a method", ident),
+                        format!("self.{}({})", path_str, args_snippet),
+                        Applicability::MachineApplicable,
+                    );
+                }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:447]: Ok, this PR adds technical debt and this whole function needs to be refactored to use a table of heuristics with priorities and flags indicating whether they support combining or not.
I'm not going to ask you to do this work, but let's try to reduce the technical debt in this PR at least.

- `smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply.
- in addition to "is a call" condition `call(self)` heuristic should also use the `(Res::Def(DefKind::Struct, def_id), _) if ns == ValueNS` condition to disable itself (this is the same condition that enables the "constructor is not visible here" heuristic in `fn smart_resolve_context_dependent_help`)
- these two heuristics should "link" to each other with comments telling that they use the same condition (but with an opposite sign)

**Comment by crlf0710** [comment]
@henryboisdequin Ping from triage. Would you mind address the comments above? Thanks!

**Comment by henryboisdequin** [comment]
@cr1901 I will address the comments right away. This PR fell off my radar for a little bit.

File: src/test/ui/codemap_tests/two_files.stderr [hunk 6]
```diff
@@ -2,13 +2,7 @@ error[E0404]: expected trait, found type alias `Bar`
...
    |
 LL | impl Bar for Baz { }
-   |      ^^^ type aliases cannot be used as traits
-   |
-help: you might have meant to use `#![feature(trait_alias)]` instead of a `type` alias
-  --> $DIR/two_files_data.rs:5:1
-   |
-LL | type Bar = dyn Foo;
-   | ^^^^^^^^^^^^^^^^^^^
+   |      ^^^ not a trait
```
Comment (henryboisdequin) [inline src/test/ui/codemap_tests/two_files.stderr:5]: Why does this cause a regression?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 7]
```diff
@@ -452,21 +469,6 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
-                source,
-                res,
-                &path_str,
-                &fallback_label,
-            ) {
-                // We do this to avoid losing a secondary span when we override the main error span.
-                self.r.add_typo_suggestion(&mut err, typo_sugg, ident_span);
-                return (err, candidates);
-            }
-        }
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:457]: This block of code should stay here, all the test changes happened because it had been changed from running unconditionally to running only for a subset of calls.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/late/diagnostics.rs:464]: >`smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply.

8b1cf540434bb55d9615f1806b966f78b2c66c6f does this incorrectly.

>in addition to "is a call" condition `call(self)` heuristic should also use the `res is Some((Res::Def(DefKind::Struct, def_id), _)) && ns == ValueNS` condition to disable itself (this is the same condition that enables the "constructor is not visible here" heuristic in `fn smart_resolve_context_dependent_help`)

8b1cf540434bb55d9615f1806b966f78b2c66c6f misses the `ns == ValueNS` condition.

>these two heuristics should "link" to each other with comments telling that they use the same condition (but with an opposite sign)

These comments were not added.

File: src/test/ui/codemap_tests/two_files.stderr (hunk 6 above)
Comment (petrochenkov) [reply src/test/ui/codemap_tests/two_files.stderr:5]: https://github.com/rust-lang/rust/pull/81950#discussion_r588858777

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 8]
```diff
@@ -428,19 +428,57 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                }
+
                 return (err, candidates);
             }
         }
 
         // Try Levenshtein algorithm.
         let typo_sugg = self.lookup_typo_candidate(path, ns, is_expected, span);
-        // Try context-dependent help if relaxed lookup didn't work.
+        // Do the same as above. These two heuristics *link* to each other and use the same condition.
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:481]: `smart_resolve_context_dependent_help` is a huge function, this comment applies only to a tiny part of it, could you move it to the relevant code?

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 9]
```diff
@@ -1844,10 +1885,13 @@ impl<'tcx> LifetimeContext<'_, 'tcx> {
...
-                                synthetic: Some(hir::SyntheticTyParamKind::ImplTrait),
-                                ..
-                            })
+                            !matches!(
+                                p.kind,
+                                hir::GenericParamKind::Type {
+                                    synthetic: Some(hir::SyntheticTyParamKind::ImplTrait),
+                                    ..
+                                }
+                            )
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:1894]: Could you remove the unrelated formatting changes?
Looks like you have some rustfmt version mismatch between your editor and `x.py fmt`.

File: compiler/rustc_resolve/src/late/diagnostics.rs [hunk 10]
```diff
@@ -428,19 +428,57 @@ impl<'a: 'ast, 'ast> LateResolutionVisitor<'a, '_, 'ast> {
...
+                            call_span,
+                            &format!("try calling `{}` as a method", ident),
+                            format!("self.{}({})", path_str, args_snippet),
+                            Applicability::MachineApplicable,
+                        );
+                    }
+                } else {
+                    err.span_suggestion(
+                        call_span,
+                        &format!("try calling `{}` as a method", ident),
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/late/diagnostics.rs:469]: There are three places reporting the "try calling `{}` as a method" suggestion, then can be deduplicated.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (petrochenkov) [reply compiler/rustc_resolve/src/late/diagnostics.rs:473]: >`smart_resolve_context_dependent_help` should be called once and we need to simply fall through to it if the `call(self)` heuristic doesn't apply

This part wasn't addressed.

**Comment by petrochenkov** [comment]
Meta: we are working with "waiting-on" labels for "passing the baton", so adding GitHub's review requests on top of them is not necessary.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@henryboisdequin can you address the review comments? Thank you.

**Comment by crlf0710** [comment]
@henryboisdequin Ping from triage. Could you address the review comments above? Thanks!

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (henryboisdequin) [reply compiler/rustc_resolve/src/late/diagnostics.rs:473]: I'm not sure how to do this as if we call `smart_resolve_context_dependent_help` at the beginning of the function, it changes the diagnostic.

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 10 above)
Comment (henryboisdequin) [reply compiler/rustc_resolve/src/late/diagnostics.rs:470]: How should I go about doing this?

Comment (henryboisdequin) [reply compiler/rustc_resolve/src/late/diagnostics.rs:470]: Ping @petrochenkov :)

File: compiler/rustc_resolve/src/late/diagnostics.rs (hunk 5 above)
Comment (henryboisdequin) [reply compiler/rustc_resolve/src/late/diagnostics.rs:474]: Ping @petrochenkov :)

**Comment by petrochenkov** [comment]
@henryboisdequin 
This is what I was trying to convey all this time - https://github.com/petrochenkov/rust/commit/455b965a816e546b1d51878b6c5df7934926cfe4?w=1

**Comment by bstrie** [comment]
@henryboisdequin , have you had time to respond to petrochenkov's most recent review?

**Comment by crlf0710** [comment]
@henryboisdequin Ping from triage, any updates on this?

**Comment by JohnCSimon** [comment]
@henryboisdequin Ping again from triage, can you please address the merge conflict?

**Comment by petrochenkov** [comment]
Closing due to inactivity.
